### DISCUSSION TITLE
Component Tracking and recursive layout

### DISCUSSION THREAD

**Discussion by DSMejantel** [sqlpage/SQLPage#313] [Q&A] [answered]
Hi !
I would like use the tracking component to watch quickly where I have to do something.
I can use a Common Table Expression with the list component like on the screenshot.
But the tracking component is more pretty and give more informations.
Il can ask him to display the 3 groups bit only the last contains all the datas :
![tracking](https://github.com/lovasoa/SQLpage/assets/149260856/722ef233-bc66-487e-8a7f-a610bfe13c51)

How can I apply the recursive expression to the top-parameters and then to the row-parameters of the component ?

This is the code : 

```sql
SET rangees= (SELECT count(distinct rang) FROM colonie where rucher_id=$id);

WITH RECURSIVE rangees(rangee,nombre) AS (
  SELECT rang, count(distinct numero) FROM colonie  WHERE colonie.rucher_id=$id and disparition::int<>1 and colonie.rang=1
  UNION ALL
  SELECT rang +1, (SELECT count(distinct numero) FROM colonie WHERE colonie.rucher_id=$id and disparition::int<>1 and colonie.rang=rangees.rangee +1)
  FROM colonie
  INNER JOIN rangees ON rangees.rangee = colonie.rang
  and rang < $rangees::int
)
select 
    'tracking'       as component,
    'rangée '||rangee||' : '||(SELECT count(distinct numero)||' ruches' FROM colonie WHERE colonie.rucher_id=$id and disparition::int<>1 and colonie.rang=rangee) as title,
    'bottom'            as placement,
    12               as width
    FROM rangees,colonie
    WHERE $tab='1' and colonie.rang=rangee GROUP BY rang
select 
    CASE WHEN colonie.tracing=2
    THEN 'warning'
    WHEN colonie.tracing=3
    THEN 'danger'
    ELSE 'success'
    END as color,
    'ruche n°'||numero||' (rangée '||rang||')' as title
    FROM colonie LEFT JOIN colvisite on colonie.numero=colvisite.ruche_id WHERE colonie.rucher_id=$id and disparition::int<>1 and $tab='1' GROUP BY colonie.numero ORDER BY rang;
```

**Comment by lovasoa** [comment]
Hi @DSMejantel !

I see you are pushing SQLPage to the edge of its capabilities. The problem in the code you are giving above is that the component selections are not at the right place. You are doing `component, component, component, data, data, data` instead of `component, data, component, data, component, data`. The latter is hard to do in sql directly, but you can use the dynamic component to achieve what you are trying to do.

[Here is a short example](https://sqliteonline.com/#sqltext=%23tab-name%3DSQLite%0D%0ADROP%20TABLE%20IF%20EXISTS%20t%3B%0Acreate%20table%20t(x%20int%2C%20y%20int)%3B%0Ainsert%20into%20t%20VALUES%20(1%2C1)%2C(2%2C1)%2C(1%2C2)%2C(2%2C2)%2C(3%2C2)%2C(1%2C3)%3B%0A%0Aselect%20'dynamic'%20as%20component%2C%0A%09'%5B'%20%7C%7C%0A%20%20%20%20%09json_object(%0A%20%20%20%20%20%20%20%20%20%20'component'%2C%20'tracking'%2C%0A%20%20%20%20%20%20%20%20%20%20'title'%2C%20'y%3D'%20%7C%7C%20y%0A%20%20%20%20%20%20%20%20)%20%7C%7C%20'%2C'%20%7C%7C%0A%20%09%09group_concat(json_object('title'%2C%20'x%3D'%20%7C%7C%20x%20))%20%7C%7C%0A%20%20%20%20'%5D'%20as%20properties%0Afrom%20t%20GROUP%20BY%20y%20ORDER%20BY%20y%3B)

```sql
select 'dynamic' as component,
  '[' ||
    json_object(
          'component', 'tracking',
          'title', 'y=' || y
    ) || ',' ||
    group_concat(json_object(
        'title', 'x=' || x
    )) ||
  ']' as properties
from t GROUP BY y ORDER BY y;
```

**Reply by lovasoa** [reply]
Here is the example running on replit: https://replit.com/@pimaj62145/SQLPage-dynamic-multi-component-example#index.sql

![image](https://github.com/lovasoa/SQLpage/assets/552629/aedba973-48a0-4c4c-bf4c-a7d66be3bf4a)

**Comment by DSMejantel** [comment, marked as answer]
Hi,
Dynamic component is magic ! Just like you ! Thanks.

```sql
select 'dynamic' as component,
  '[' ||
    json_object(
          'component', 'tracking',
          'title', 'Rangée ' || rang,
          'width', 4
    ) || ',' ||
    group_concat(json_object(
        'title', 'Ruche '||numero,
        'color', case tracing when 3 then 'red' when 2 then 'orange' else 'green' end
    )) ||
  ']' as properties
from colonie WHERE colonie.rucher_id=$id and disparition::int<>1 and $tab='1' GROUP BY rang ORDER BY rang;
```

![tracking2](https://github.com/lovasoa/SQLpage/assets/149260856/a4d4645e-0323-40c9-9052-f5113fd8dbb3)
